src/System.Private.ServiceModel/src/System/ServiceModel/Channels/CommunicationObject.cs

```diff
@@ -758,14 +767,48 @@ internal void ThrowPending()
 
         internal protected virtual Task OnCloseAsync(TimeSpan timeout)
         {
-            Contract.Requires(false, "OnCloseAsync needs to be implemented on derived classes");
-            return TaskHelpers.CompletedTask();
+            // Derived classes are required to override this method because we
+            // use these Task-based API's exclusively with the product.
+            return Task.FromException(new InvalidOperationException(
```

Comment:
You could just throw. The difference would be functionally equivalent to the difference between an exception being thrown before or after an await which completed asynchronously. I don't think it matters much as we shouldn't be hitting this code anyway and the exception message will tell you exactly where it came from.